packages/origin-247-certificate/src/offchain-certificate/synchronize/handlers/issue-persist.handler.ts

```diff
@@ -0,0 +1,119 @@
+                blockchainCertificateId: certificate.id
+            });
+            return { success: true };
+        } else {
+            await this.offchainCertificateService.persistError(event.internalCertificateId, {
+                errorMessage: `Cannot issue certificate for certificate id: ${event.internalCertificateId}`
+            });
+            return { success: false };
+        }
+    }
+
+    async synchronizeBatchBlock(
+        events: CertificateEventEntity[]
+    ): Promise<{ failedCertificateIds: number[] }> {
+        const issuedEvents = events as CertificateIssuedEvent<null>[];
+
+        const certificateIds = await this.certificateService.batchIssue(
+            issuedEvents
+                .map((issuedEvent) => issuedEvent.payload)
+                .map((payload) => ({
+                    ...payload,
+                    fromTime: new Date(payload.fromTime),
+                    toTime: new Date(payload.toTime)
+                }))
+        );
+
+        const failedCertificateIds = await Promise.all(
+            events.map(async (event, index) => {
+                const areCertificateIdsValid =
+                    certificateIds &&
+                    certificateIds.every((certificateId) =>
+                        IssuePersistHandler.isCertificateIdValid(certificateId)
+                    );
+
+                if (areCertificateIdsValid) {
+                    await this.offchainCertificateService.issuePersisted(
+                        event.internalCertificateId,
+                        {
+                            blockchainCertificateId: certificateIds[index]
+                        }
+                    );
+                } else {
+                    await this.offchainCertificateService.persistError(
+                        event.internalCertificateId,
+                        {
+                            errorMessage: `Cannot issue certificate for certificate id: ${event.internalCertificateId}`
+                        }
+                    );
+                    return event.internalCertificateId;
+                }
+            })
+        );
+
+        return { failedCertificateIds: compact(failedCertificateIds) };
+    }
+
+    async handleBatch(events: CertificateEventEntity[]) {
+        const eventsBlocks = chunk(events, this.batchConfiguration.issueBatchSize);
+        const failedCertificateIds: number[] = [];
+
+        for (const eventsBlock of eventsBlocks) {
+            const nonFailedEvents = eventsBlock.filter(
+                (event) => !failedCertificateIds.includes(event.internalCertificateId)
+            );
+
+            const result = await this.synchronizeBatchBlock(nonFailedEvents);
+            failedCertificateIds.push(...result.failedCertificateIds);
+        }
+
+        return { failedCertificateIds };
+    }
+
+    private static isCertificateIdValid(certificateId: number): boolean {
```

Comment:
why is is this necessary?

Reply:
I guess that `certificate.id === 0` is valid so I cannot check `if(certificate.id) { persist() }` so I'm just checking if it's integer